source/slang/slang-ir-specialize-function-call.cpp

```diff
@@ -762,6 +762,44 @@ struct FunctionParameterSpecializationContext
             oldFunc,
             newFunc);
 
+
+        // If we have added an Linkage decoration, we want to remove and destroy it,
+        // because the linkage should only be on the original function and
+        // not on the "torn off" copies made in this function.
+        //
+        // It *could* be argued that we don't want to duplicate the decoration instructions
+        // to begin with, just to throw them away. That may be true, but it's simpler to just remove
+        // them than filter out in cloning.
+
+        {
+            auto decorationList = newFunc->getDecorations();
+
+            const auto end = decorationList.end();
+            auto cur = decorationList.begin();
+
+            while(cur != end)
+            {
+                IRDecoration* decoration = *cur;
+
+                // We step before before the test/destroying to ensure cur is not pointing
+                // to a now destroyed instruction
+                ++cur;
+
+                if (as<IRLinkageDecoration>(decoration))
+                {   
+                    decoration->removeFromParent();
+                    decoration->removeArguments();
+
+                    // We DON'T deallocate the children/decorations, as they will be used on the original
+                    // This is okay as it stands, because linkage decorations just contain a string literal  
+
+                    // Run destructor to be sure...
+                    decoration->~IRDecoration();
```

Comment:
I *think* it would be safe to just use `decoration->removeAndDeallocate()` here.

The comment here seems to be confused about the distinction between operands and children/decorations:

* An instruction does not own its operands, because they could be shared with other instructions (the use/def relationship forms a directed graph - hopefully a DAG). The `removeAndDeallocate()` operation unlinks an insruction as a user of its operands, but does not otherwise try to delete/remove the instructions that are used as oprands.

* An instruction full owns its children (and decorations and parameters are just special cases of children), and they cannot be shared with other instructions (the parent/child relationship is always a tree). While the children of a parent instruction P might be used as operands of other instructions, those other instructions must also be descendents of P given our SSA formulaton.

Instructions should pretty much always be removed using `removeInstAndDeallocate()`.

Reply:
Okay fixed this, thanks for the clarification.